File: pp-exercises/mpi/exercise2/ring-reduction/ring-reduction.f90
program main  ! global and partial sum

  implicit none
  include 'mpif.h'
  integer :: size, rank, ierr
  integer :: sum, i

  call mpi_init(ierr)
  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)

  call mpi_reduce(rank, sum, 1, MPI_INTEGER, MPI_SUM, 0, MPI_COMM_WORLD, ierr)

  if (rank == 0) write(6,*) "global sum = ", sum

  call mpi_scan(rank, sum, 1, MPI_INTEGER, MPI_SUM, MPI_COMM_WORLD, ierr)

  if (rank /= 0) then
     call mpi_send(sum, 1, MPI_INTEGER, 0, 0, MPI_COMM_WORLD, ierr)
  else
     do i = 0, size - 1
        if (i > 0) then
           call mpi_recv(sum, 1, MPI_INTEGER, i, 0, &
                         MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
        endif
        write(6,*) " partial sum(", i, ") = ", sum
     enddo
  endif

  call mpi_finalize(ierr)

end

File: pp-exercises/mpi/exercise2/ring-reduction/ring-reduction.c
#include <stdio.h>
#include <mpi.h>

int main(int argc, char *argv[])  // global and partial sum
{
    int size, rank;
    int sum, i;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    MPI_Reduce(&rank, &sum, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);

    if (rank == 0) printf("global sum = %d\n", sum);

    MPI_Scan(&rank, &sum, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);

    if (rank != 0) {
	MPI_Send(&sum, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
    } else {
	for (i = 0; i < size; i++) {
	    if (i > 0) MPI_Recv(&sum, 1, MPI_INT, i, 0, 
				MPI_COMM_WORLD, MPI_STATUS_IGNORE);
	    printf("partial sum(%d) = %d\n", i, sum);
	}
    }

    MPI_Finalize();
    return 0;
}  
